### ISSUE TITLE
Get rid of `eastl::begin()`, et. al.?

### ISSUE THREAD

**Issue by jwdevel** [electronicarts/EASTL#356]
I suppose this is mostly a question, rather than a bug report.

**Motivation:** when using [Catch2](https://github.com/catchorg/Catch2/) with EASTL, I ran into a compile error ([link](https://github.com/catchorg/Catch2/issues/1882)).
Now, that error might be a fault of Catch2, but regardless, it started some investigation into `eastl::begin()` and related.

Note: I'll discuss `begin()`, but this really applies to (almost?) everything guarded by `EASTL_BEGIN_END_ENABLED`.

Judging by [the comments in iterator.h](https://github.com/electronicarts/EASTL/blame/38e2811b5097bd0fefb7a5379e8b661591f7bc26/include/EASTL/iterator.h#L1046), the code was added to support range-based for loops.

However, as far as I can tell, these functions are not needed for that purpose.
The continer classes already have member `.begin()` and related methods, and range-based for loops will use those preferentially.

In fact, AFAIK, even `std::begin` will defer to the member method, if it is present.

Basically, I don't see that the `eastl::` functions add anything over the corresponding `std::` functions, and (as exemplefied by the compile error mentioned above), they can result in ambiguity.

I the issue fixed for myself by setting `#define EASTL_BEGIN_END_ENABLED 0` with no ill effects so far.
But maybe there is some explanation I am missing...

**Comment by rparolin** [comment]
I'm not sure I follow because these implementations are mandated by standard specification.
https://en.cppreference.com/w/cpp/iterator/begin

We don't defer to the std:: implementation here so we provide our own.  I hope this helps.

**Comment by jwdevel** [comment]
> We don't defer to the std:: implementation here...

Yes, I guess that's my question: why doesn't EASTL defer to the std:: implementation?
As detailed a bit more in that [Catch2 link](https://github.com/catchorg/Catch2/issues/1882), the problem seems to occur because of ambiguity between `std::begin` and `eastl::begin`.

Are you saying that EASTL maintains equivalent versions of the `std::` ones out of principle?
They don't seem to provide any extra functionality, given that `std::begin` already automatically defers to the per-member `container.begin()`, which EASTL defines.

So in my mind, `eastl::begin` adds no benefit, and has some cost (see my Catch2 issue).
Hence my wondering if it can be removed (:
